Manage AO segments independently in QE nodes

hlinnaka opened the pull request:
This is a rebase of https://github.com/greenplum-db/gpdb/pull/2242, which in turn was a rebase of https://github.com/greenplum-db/gpdb/pull/790.

There are some regression failures still to fix, and I probably broke some stuff with this latest rebase. But I'd love to get this in.

I'm not sure what the remaining issues with PR #2242 were. The pipeline must be made green, but other than that, does anyone remember if there was some problem that was not evident in the pipeline?

tvesely commented on src/backend/access/aocs/aocssegfiles.c:
```diff
@@ -116,6 +120,25 @@ InsertInitialAOCSFileSegInfo(Relation prel, int32 segno, int32 nvp)
+	result = heap_lock_tuple(segrel, segtup,
+							 GetCurrentCommandId(true),
+							 LockTupleExclusive,
+							 LockWaitSkip,
```
I don't think `LockWaitSkip` makes sense in this case? When we are inserting shouldn't this be `LockWaitError` or similar?

hlinnaka replied:
Yeah, locking the tuple really should succeed here. It was that way in after the last rebase, if I read it correctly, I just kept this rebase faithful to the original for now. I suspect there might be a race condition here, anyway, hence I added the FIXME.

tvesely commented on src/backend/access/appendonly/aosegfiles.c:
```diff
@@ -143,8 +150,26 @@ InsertInitialSegnoEntry(Relation parentrel, int segno)
+	result = heap_lock_tuple(pg_aoseg_rel, pg_aoseg_tuple,
+							 GetCurrentCommandId(true),
+							 LockTupleExclusive,
+							 LockWaitSkip,
```
Same as with `InsertInitialAOCSFileSegInfo`, shouldn't this be `LockWaitError`?

hlinnaka commented:
Prior discussion on the vacuum and visibility issues: https://groups.google.com/a/greenplum.org/d/msg/gpdb-dev/49MW3tBoH8s/FzL3ascmAgAJ

asimrp commented on src/backend/commands/vacuum_ao.c:
```diff
@@ -0,0 +1,834 @@
+	 * (Problem 3), so we would need to coordinate the transactions from the
+	 * QD.
+	 *
+	 * FIXME: Imagine that there are two segfiles with
```
This seems like a preexisting bug.  Should `SetSegnoInternal()` be changed to choose an unused segno as target for compaction?  Keep using this segno until it fills up to the threshold of 90%.

hlinnaka replied:
I changed the preferences of selecting insertion target for VACUUM compaction to:

1. Reuse an existing empty segfile, if any
2. Create a new segfile
3. Reuse an existing non-empty segfile.

That should fix take care of problem in most cases. You could still run into it if all segfiles are in use, so a new one cannot be created, but that should be rare.

hlinnaka commented on src/backend/access/appendonly/aosegfiles.c:
```diff
@@ -143,8 +150,26 @@ InsertInitialSegnoEntry(Relation parentrel, int segno)
+	 * Lock the tuple so that concurrent insert transaction will not
+	 * consider this segfile for insertion.
+	 *
+	 * FIXME: Is there still a race condition, if another backend locks
```
Hmm. Then the entry would go away if the transaction aborts, so a vacuum would not know to truncate the file.

The bigger race condition here, actually, is that the callers are basically doing this:

1. Check if the pg_aoseg entry for this segfile exists
2. Create it if not.

We need some lock to avoid the scenario that two backends determine that the segfile doesn't exist at the same time, and then both try to create it. There are two callers to this function,  SetSegnoInternal() and SetCurrentFileSegForWrite(). SetSegnoInternal() grabs a file-extension lock on the table, to prevent other concurrent calls to SetSegnoInternal(), and SetCurrentFileSegForWrite() grabs a lock on the AO segfile with LockRelationAppendOnlySegmentFile(). They're not using the same lock, so they're not safe against each other. However, all callers of SetCurrentFileSegForWrite() are either commands like ALTER TABLE that hold an AccessExclusiveLock on the table and use segfile 0, or they called SetSegnoInternal() (via SetSegnoForWrite()) which should have created the pg_aoseg entry already.

So I think this is OK as it is, but it could use some cleanup...

hlinnaka commented:
Thanks for the review, @asimrp! I'm continuing to clean this up, I think this is functionally pretty much ready, but the code is a bit of a mess.

I pushed a few new commits to clean up the way the locking works. I realized that we don't really need the heavy-weight locks on the AO segments, after all. All operations that modify a segfile, or its pg_aoseg entry, need to lock the pg_aoseg tuple. I cleaned up a few places where were not very clear on the locking, and removed the heavy-weight lock as redundant. I added also some notes to the README on how the locking is supposed to work.

I haven't run the full regression suite with these latest changes, so I expect some new test failures. I'll slog through them once concourse has finished.

hlinnaka commented:
> And by performing compaction within a distributed transaction, all the problems with previous versions of this patch seem to have been resolved.

Should note one new issue with doing the compaction in a distributed transaction, though: we now do compact *all* the segfiles of a table in one big distributed transaction. Recycling the AWAITING_DROP segfiles doesn't happen until all the segfiles have been compacted. Previously, the QD would issue a separate transaction for each segfile. One big transaction can require a lot more disk space, if the table has a lot of segfiles that require compaction.

It'd be nice to do something about that, but it would make the dispatching a lot more complicated. Maybe later...

hlinnaka commented on src/backend/access/appendonly/aosegfiles.c:
```diff
@@ -1244,66 +1375,7 @@ gp_update_aorow_master_stats_internal(Relation parentrel, Snapshot appendOnlyMet
-				 * only expect this function to update segno 0 only and the QD
-				 * never cares about segment 0 anyway.
-				 */
+				/* FIXME: This does nothing now.  Do we need this function at all? */
```
I assumed "no", and removed it.

jimmyyih replied:
The gpcrondump tool used it but apparently not the new gpbackup... so sure, let's remove it.

hlinnaka replied:
> The gpcrondump tool used it but apparently not the new gpbackup... so sure, let's remove it.

Are you sure? I can see some references to "modcount" in the gpbackup repository.

jimmyyih replied:
Ahh, okay I looked a bit deeper into this since I just blindly grepped in the gpbackup repository and its dependencies without looking at what the `gp_update_ao_master_stats()` was being used for.  I thought it was more related to the dump side sorry.

So technically, the new `gprestore` utility (not to be confused with `gp_restore/cdb_restore`) is not updating the corresponding `pg_aoseg` entry accordingly for tupcount and modcount... which is not good.  It can be very easy to blame the `gprestore` utility not calling `gp_update_ao_master_stats()`... but IMO, the `COPY FROM SEGMENT` should be updating the `pg_aoseg` entry automatically at end of transaction just like regular `COPY`.  Maybe we could call `UpdateMasterAosegTotals()` at the end of `CopyDispatchOnSegment()`?  I'll play around with it and either open up a PR or an Issue later this week.

But coming back to this particular PR, yeah incremental AO backup depends on modcount from the master `pg_aoseg` table which will no longer be populated... but this was already discussed in the first iteration of this PR in https://github.com/greenplum-db/gpdb/pull/790#issuecomment-239639483.  Should be fine to change the logic to use `gp_dist_random()` or the gp_toolkit helper functions (I'll take care of it on the gpbackup repository).  But as far as the function in this comment thread goes, `gp_update_ao_master_stats()` is indeed a dead function and can be removed.

chrishajas replied:
analyzedb also uses the same mechanism as backup to identify ao tables that have been modified since last analyze.

hlinnaka commented on src/backend/commands/vacuum_ao.c:
```diff
@@ -0,0 +1,830 @@
+	Snapshot	appendOnlyMetaDataSnapshot = RegisterSnapshot(GetCatalogSnapshot(InvalidOid));
+
+	/*
+	 * This runs in a *distributed* transaction:
```
Added.

hlinnaka commented:
Woohoo, the PR pipeline is now green! I'll do a little bit more tidying up on this, but this is basically ready for commit now. Please review!

@asimrp, you looked at this earlier, but some things have changed since your last review. The commits in this PR need to be squashed before pushing, but you might want to take a look at the commits from last few days. In particular, commit "Clarify locking.", and the changes after that.

asimrp commented on src/backend/access/aocs/aocssegfiles.c:
```diff
@@ -163,7 +163,8 @@ InsertInitialAOCSFileSegInfo(Relation prel, int32 segno, int32 nvp)
 				   Snapshot appendOnlyMetaDataSnapshot,
-				   int32 segno)
+				   int32 segno,
+				   bool locked)
```
Let's update the comment above this function and include a description of `locked` in it.

asimrp commented on src/backend/access/aocs/aocssegfiles.c:
```diff
@@ -549,20 +561,6 @@ SetAOCSFileSegInfoState(Relation prel,
-													  segno,
-													  AccessExclusiveLock,
-													   /* dontWait */ false);
-	if (acquireResult != LOCKACQUIRE_ALREADY_HELD)
```
The obsolete comment on lines 568-569 needs to be removed.

asimrp commented on src/backend/access/appendonly/aosegfiles.c:
```diff
@@ -587,8 +602,6 @@ ClearFileSegInfo(Relation parentrel,
 				 int segno,
 				 FileSegInfoState newState)
 {
-	LockAcquireResult acquireResult;
```
This is the code path to drop AWAITING_DROP segments, so we don't expect the tuple to be already locked.  Should we verify that the tuple is NOT already locked, in that case?

hlinnaka replied:
Hmm. There's no harm in locking the row first, even if it's unnecessary here, so I'd rather not check for that. But I'll try to make the comments more clear.

hlinnaka commented:
I rebased this again, and squashed all the commits into one final commit, and rewrote the commit message. Please review the commit message.

Barring any last-minute comments or pipeline failures, this is ready! I plan to push this in the next few working days.

hlinnaka commented:
Pushed! Thank you everyone!
